### ISSUE TITLE
Truncated Cloudwatch log message

### ISSUE THREAD

**Issue by robertlarkins** [Cimpress-MCP/serilog-sinks-awscloudwatch#97]
We're using Serilog.Sinks.AwsCloudWatch v4.0.159 to log from a Asp.Net core app running in Fargate (on a linux container) to CloudWatch logs. However, long messages (such as logging Reqests) are being truncated at about 3000 characters, well, I say truncated, the rest of the message seems to be an ASCII nul character. My expectation is that CloudWatch should be able to hold events up to 256KB.

I'm really not sure where or why this might be occurring. This is the `CloudWatchSinkOptions` we're have:
```C#
var awsLoggerConfig = new CloudWatchSinkOptions
{
    LogGroupName = logGroupName,
    LogStreamNameProvider = new ConfigurableLogStreamNameProvider(logStreamName, appendHostName: false, appendUniqueInstanceGuid: false),
    TextFormatter = new CompactJsonFormatter(),
    CreateLogGroup = true,
    MinimumLogEventLevel = LogEventLevel.Verbose,
    BatchSizeLimit = 100,
    QueueSizeLimit = 10000,
    Period = TimeSpan.FromSeconds(10),
    RetryAttempts = 5
};

var cloudWatchClient = new AmazonCloudWatchLogsClient(awsConfiguration.RegionEndpoint);

return new LoggerConfiguration()
    .MinimumLevel.Override("Microsoft.AspNetCore", LogEventLevel.Warning)
    .Enrich.FromLogContext()
    .WriteTo.Console()
    .WriteTo.AmazonCloudWatch(awsLoggerConfig, cloudWatchClient)
    .CreateLogger();
```

and this is what part of the log message looks like:
> 2020-09-21 03:10:19.577,"[03:10:19 INF] Request: http *[correct data for ~3000 characters, then]* Code"":""DC"",""                                  *[then about 94,000 nul characters]*                                                          "

All up the message should be about 98,000 characters long, but diagnosing issues with partial messages is problematic. Any ideas if this is a setup issue, a limitation somewhere (seems strange if it is still putting out the nul characters), or potentially a bug somewhere?

**Comment by robertlarkins** [comment]
Looking further, other logging places seem to be providing the whole message. The logging that is getting truncated comes from some middleware, maybe the issue is in here somewhere:
```C#
    /// <summary>
    /// Middleware for logging the raw request or response.
    /// This comes from: https://exceptionnotfound.net/using-middleware-to-log-requests-and-responses-in-asp-net-core/
    /// Or from this GitHub Repo: https://github.com/exceptionnotfound/AspNetCoreRequestResponseMiddlewareDemo .
    /// </summary>
    public class RequestResponseLoggingMiddleware
    {
        private readonly RequestDelegate next;

        public RequestResponseLoggingMiddleware(RequestDelegate next)
        {
            this.next = next;
        }

        public async Task Invoke(HttpContext context)
        {
            // First, get the incoming request
            var request = await FormatRequest(context.Request);

            // Copy a pointer to the original response body stream
            var originalBodyStream = context.Response.Body;

            // Create a new memory stream...
            using var responseBody = new MemoryStream();

            // ...and use that for the temporary response body
            context.Response.Body = responseBody;

            // Continue down the Middleware pipeline, eventually returning to this class
            await next(context);

            // Format the response from the server
            var response = await FormatResponse(context.Response);

            Log.Logger.Information($"Request: {request}");
            Log.Logger.Information($"Response: {response}");

            // Copy the contents of the new memory stream (which contains the response) to the original stream, which is then returned to the client.
            await responseBody.CopyToAsync(originalBodyStream);
        }

        private async Task<string> FormatRequest(HttpRequest request)
        {
            // This line allows us to set the reader for the request back at the beginning of its stream.
            request.EnableBuffering();

            // We now need to read the request stream.  First, we create a new byte[] with the same length as the request stream...
            var buffer = new byte[Convert.ToInt32(request.ContentLength)];

            // ...Then we copy the entire request stream into the new buffer.
            await request.Body.ReadAsync(buffer, 0, buffer.Length);

            // We convert the byte[] into a string using UTF8 encoding...
            var bodyAsText = Encoding.UTF8.GetString(buffer);

            // Reset the stream to start, which is allowed because of EnableBuffering()
            request.Body.Seek(0, SeekOrigin.Begin);

            return
                $"{request.Scheme} {request.Host}{request.Path} {request.QueryString} {bodyAsText}";
        }

        private async Task<string> FormatResponse(HttpResponse response)
        {
            // We need to read the response stream from the beginning...
            response.Body.Seek(0, SeekOrigin.Begin);

            // ...and copy it into a string
            string text = await new StreamReader(response.Body).ReadToEndAsync();

            // We need to reset the reader for the response so that the client can read it.
            response.Body.Seek(0, SeekOrigin.Begin);

            // Return the string for the response, including the status code (e.g. 200, 404, 401, etc.)
            return $"{response.StatusCode}: {text}";
        }
    }
```

**Comment by thoean** [comment]
Does your last comment mean it's not related to this library, and that you're finding the problem in your code? Thanks for adding the right context to ensure I'm not missing the point.

**Comment by robertlarkins** [comment]
@thoean There is still an issue occurring as when I run this locally the log has the full request message (no truncation). The setup of the Logger for both local and CloudWatch is done here, where if running on local the `if (awsConfiguration.IsLocalEnvironment)` path is taken:

```C#
public class LoggerConfig : ILoggerConfig
{
    public ILogger CreateLogger(IAwsConfiguration awsConfiguration)
    {
        var env = awsConfiguration.EnvironmentName;

        if (awsConfiguration.IsLocalEnvironment)
        {
            var currentDate = DateTimeOffset.Now.ToString("ddMMyyyyhhmmss");
            var logFile = $"{Directory.GetCurrentDirectory()}/logs/log_run_{currentDate}.log";

            return new LoggerConfiguration()
                .Enrich
                .FromLogContext()
                .WriteTo.Console()
                .WriteTo.File(logFile)
                .CreateLogger();
        }

        var logGroupName = $"/{env}/logs";
        var logStreamName = "api";

        Console.WriteLine("Logs :: Setting up CloudWatch Logs");
        Console.WriteLine($"Logs :: Log Group: {logGroupName}");
        Console.WriteLine($"Logs :: Log Stream: {logStreamName}");

        var awsLoggerConfig = new CloudWatchSinkOptions
        {
            LogGroupName = logGroupName,
            LogStreamNameProvider = new ConfigurableLogStreamNameProvider(logStreamName, appendHostName: false, appendUniqueInstanceGuid: false),
            TextFormatter = new CompactJsonFormatter(),
            CreateLogGroup = true,
            MinimumLogEventLevel = LogEventLevel.Verbose,
            BatchSizeLimit = 100,
            QueueSizeLimit = 10000,
            Period = TimeSpan.FromSeconds(10),
            RetryAttempts = 5
        };

        var cloudWatchClient = new AmazonCloudWatchLogsClient(awsConfiguration.RegionEndpoint);

        return new LoggerConfiguration()
            .MinimumLevel.Override("Microsoft.AspNetCore", LogEventLevel.Warning)
            .Enrich.FromLogContext()
            .WriteTo.Console()
            .WriteTo.AmazonCloudWatch(awsLoggerConfig, cloudWatchClient)
            .CreateLogger();
    }
}
```

**Comment by thoean** [comment]
Interesting... UTF-8 should be a max of 4 bytes per character, and your message should be below the threshold (or at least not cut off that much).

Here's the code that trims the message length: https://github.com/Cimpress-MCP/serilog-sinks-awscloudwatch/blob/master/src/Serilog.Sinks.AwsCloudWatch/CloudWatchLogSink.cs#L383

Would you mind [enabling debug log](https://github.com/serilog/serilog/wiki/Debugging-and-Diagnostics) and see if you run into this code path? And as a follow up, if you could capture that message before trimming, what happens to the next line referenced above (where a substring is done)?

**Comment by robertlarkins** [comment]
@thoean Thursday will be the earliest chance I'll get to get to run those logs, will report back my findings once done.

**Comment by robertlarkins** [comment]
added the following line to the CreateLogger method: `SelfLog.Enable(msg => Console.WriteLine("SeriLog CloudWatchSinks: " + msg));` just before creating `CloudWatchSinkOptions`.
The request body was 60,298 characters long (not sure how to output the request other than logging once in AWS). This got truncated after about ~59,400 characters. But there was no debug log message captured from serilog for "Truncating log event with length of {0}". So I can only assume it wasn't hit. Other logs were captured from the serilog debugging such as:
| @message | SeriLog CloudWatchSinks: 2020-10-02T03:31:36.0139390Z Invalid sequence token. Attempt: 0 Error: Amazon.CloudWatchLogs.Model.InvalidSequenceTokenException: The given sequenceToken is invalid. The next expected sequenceToken is: 49605620959753294569043380895932015498833336604878217106
-- | --
@timestamp | 1601609496015

and

@message | SeriLog CloudWatchSinks: 2020-10-02T03:31:38.3322336Z Required properties not provided for: Response: 200: {
-- | --
@timestamp | 1601609498332

**Comment by thoean** [comment]
Ok, so we didn't hit that. From that aspect, there _should_ be nothing else in this library that shortens the message.

What text formatter are you using? Is there an option to debug that? Alternatively, could you download the source code here, add it as a direct dependency instead of the library dependency, set a breakpoint and try it out locally (changing the local configuration to send to CloudWatch instead of the local version of it)?

In short, any options you could have a debug in the file I mentioned before, which will hopefully give the right hints whether the textformatter is doing something special, or something we missed in this library but wasn't captured in the debug log, etc.

Thank you for your ongoing efforts to help find and eventually eliminate this problem.

**Comment by robertlarkins** [comment]
So after all that it was an issue in the request logging middleware. Just took a while to find it as worked from the wrong end... Anyway, thanks for your help.

**Comment by thoean** [comment]
I'm glad you figured it out and thanks for taking the extra loop to look into the details of this library.
